matter_server/server/device_controller.py

```diff
@@ -0,0 +1,391 @@
+        ClusterEvent,
+        ClusterCommand
+    )
+
+    from .server import MatterServer
+    from .stack import MatterStack
+
+
+DATA_KEY_NODES = "nodes"
+DATA_KEY_LAST_NODEID = "last_nodeid"
+
+
+class MatterDeviceController:
+    """Class that manages the Matter devices."""
+
+    def __init__(
+        self,
+        server: MatterServer,
+    ):
+        self.server = server
+        self.logger = server.logger.getChild("device_controller")
+        # Instantiate the underlying ChipDeviceController instance on the Fabric
+        self.chip_controller: ChipDeviceController = (
+            server.stack.fabric_admin.NewController()
+        )
+        # we keep the last events in memory so we can include them in the diagnostics dump
+        self.event_history: Deque[Attribute.EventReadResult] = deque(maxlen=25)
+        self._subscriptions: dict[int, Attribute.SubscriptionTransaction] = {}
+        self._nodes: dict[int, MatterNode] = {}
+        self._wifi_creds_set = False
+        self.logger.debug("CHIP Device Controller Initialized")
+
+    @property
+    def fabric_id(self) -> int:
+        """Return Fabric ID."""
+        return self.chip_controller.fabricId
+
+    @property
+    def compressed_fabric_id(self) -> int:
+        """Return unique identifier for this initialized fabric."""
+        return self.chip_controller.GetCompressedFabricId()
+
+    async def start(self) -> None:
+        """Async initialize of controller."""
+        # load nodes from persistent storage
+        nodes_data = self.server.storage.get(DATA_KEY_NODES, {})
+        for nodeid_str, node_dict in nodes_data.items():
+            nodeid = int(nodeid_str)
+            # TEMP !!!! TODO
+            # node_dict["attributes"] = {}
+            node = dataclass_from_dict(MatterNode, node_dict)
+            self._nodes[nodeid] = node
+            # make sure to start node subscriptions
+            await self._subscribe_node(nodeid)
+        # create task to check for nodes that need any re(interviews)
+        self.server.loop.create_task(self._check_interviews())
+        self.logger.debug("Started.")
+
+    async def stop(self) -> None:
+        """ "Handle logic on server stop."""
+        # unsubscribe all node subscriptions
+        for sub in self._subscriptions.values():
+            await self._call_sdk(sub.Shutdown)
+        self._subscriptions = {}
+        await self._call_sdk(self.chip_controller.Shutdown)
+        self.logger.debug("Stopped.")
+
+    @api_command("device_controller.GetNodes")
+    def get_nodes(self) -> list[MatterNode]:
+        """Return all Nodes known to the server."""
+        return [x for x in self._nodes.values() if x is not None]
+
+    @api_command("device_controller.GetNode")
+    def get_node(self, nodeid: int) -> MatterNode:
+        """Return info of a single Node."""
+        node = self._nodes.get(nodeid)
+        assert node is not None, "Node does not exist or is not yet interviewed"
```

Comment:
This is just type annotation assertion or check for developers, right? If it can happen in production we shouldn't use `assert`.

Reply:
yeah, used it while testing. To be removed later